django/core/exceptions.py

```diff
@@ -182,6 +184,30 @@ def __str__(self):
     def __repr__(self):
         return 'ValidationError(%s)' % self
 
+    def __eq__(self, other):
+        if not isinstance(other, ValidationError):
+            return NotImplemented
+        if hasattr(self, 'message'):
+            return (
+                hasattr(other, 'message') and
+                self.message == other.message and
+                self.code == other.code and
+                self.params == other.params
+            )   
+        return (
+            hasattr(self, 'error_list') == hasattr(other, 'error_list') and
+            hasattr(self, 'error_dict') == hasattr(other, 'error_dict') and
+            sorted(self.messages) == sorted(other.messages)
+        )
+
+    def __hash__(self):
+        if hasattr(self, 'message'):
+            return hash((make_hashable(self.message), make_hashable(self.code), make_hashable(self.params)))
+        if hasattr(self, 'error_list'):
+            return hash((make_hashable(sorted(self.messages)), 'error_list'))
+        if hasattr(self, 'error_dict'):
+            return hash((make_hashable(sorted(self.messages)), 'error_dict'))
```

Comment:
I changed `make_hashable` to sort `dicts` as I think that is for the neater solution. For lists it doesn't matter as much so I left the sorting here, but it could also go to `make_hashable`?